File: Data Structures & Algorithms/binary-tree-postorder-traversal/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        lst = []
        stack1 = []
        stack2 = []

        if root is None:
            return lst
        
        stack1.append(root)

        while stack1:
            root = stack1.pop()
            stack2.append(root)
            if root.left:
                stack1.append(root.left)
            if root.right:
                stack1.append(root.right)
        
        while stack2:
            root = stack2.pop()
            lst.append(root.val)

        return lst

File: Data Structures & Algorithms/lowest-common-ancestor-of-a-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    unordered_map<TreeNode*, pair<TreeNode*, int>>mp;
public:
    void helper(TreeNode* root, TreeNode* parent, int level){
        if(root){
            mp[root] = {parent, level};
            helper(root->left, root, level+1);
            helper(root->right, root, level+1);
        }
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        mp.clear();
        helper(root, root, 0);
        auto p1 = mp[p];
        auto q1 = mp[q];

        TreeNode* first = p, *second = q;
        int d1 = p1.second, d2 = q1.second;

        if(d2 > d1){
            swap(first, second);
        }
        int diff = abs(d1-d2);

        while(diff!=0){
            first = mp[first].first;
            diff --;
        }

        while(first != second){
            first = mp[first].first;
            second = mp[second].first;
        }
        return first;
    }
};